README.md

```diff
@@ -1,9 +1,166 @@
+
+```
+python generator/generator.py
+```
+
+This will parse the test specs in generator/testspecs and create test data in testdata/.  
+
+# Test Data
+The generator/generator.py script creates test data files in the testdata folder.  These test data files are intended for use, either with the mock service supplied here, or in other mock services that may be developed for the parsec service or parsec clients.  The test data files consist of the [test spec](#test-spec-format) (useful for the writer of a unit test), and test data, which contains base 64 encodes strings for a parsec request and a corresponding result.
+
+The overall test data file format is, therefore:
+
+```yaml
+spec:
+    # for format see below
+test_data:
+  request: EKfAXh4AAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAA
+  result: EKfAXh4AAQAAAAAAAAAAAAAAAAAAAAIAAAAAAAEAAAAAAAAACAE=
+```
+## Expected Use Of Test Data
+This repository does not dictate how the test data should be used by parsec client developers, but it was developed with the following use case in mind:
+
+A client developer wants to test a parsec request in their client.  
+- They select an appropriate test data file that fulfils the needs of their test (or creates a new one in this repository and uses that)
+- They write client code that should generate a message to match the request section of the [test spec](#test-spec-format).
+- They configure a mock service (the one here or a custom one) to expect the data defined in the test_data.request part of the test data file as well as the data to send if that request is received and what to send if it is not.
+- They stimulate their client under test to send the request message to the mock service.
+- The mock service will check to see if the request message it received matches the expected value (this is an exact byte match).
+  - If the match was successful, then it will return the configured match response
+  - Otherwise, it will return the configured non-match response
+- The client under test will attempt to decode the response.  The results can be checked by the test code.
+- If the spec.response.parse_should_succeed field is set to true, then the test code *should* expect a successful parse of the response data.  Otherwise, it *should* expect the parsing to fail.
+
+
+## Writing Tests
+To write a new test, a developer should create a test spec file in the generator/testspecs folder, using the [format below](#test-spec-format). 
+
+They then need to write a generator function in ```generator/generator_lib.py```.  This generator function takes no arguments, and should output a tuple of python binary strings.  
+
+The first element of the tuple should be the protocol buffer encoded value of the request message that is described in the spec.request.body_description field of the test spec.  The second element of the tuple should be the protocol buffer encoded value of the result message that is described in the spec.result.body_description field of the test spec.
```

Comment:
I think it would be good to align on terminology in this whole PR. I think [from the book](https://parallaxsecond.github.io/parsec-book/parsec_client/wire_protocol.html) there are multiple different things:
- a *request* which is the whole buffer of data sent to the service containing:
  - the fixed common header
  - the message body bytes: the *operation* (serialised protobuf)
  - the authentication bytes
- a *response* which is the whole buffer sent back from the service containing:
  - the fixed common header
  - the message body bytes: the *result* (serialised protobuf)
  
My point being that we should be careful when we use the terms `request`, `response`, `operation`, `result`.

In this case, the `gen_` methods really return the serialised "operation" and "request" as opposed as "request" and "response".

Reply:
wilco

Reply:
fixed, in here and in code, i think.

Reply:
thanks!